python/cudf_polars/cudf_polars/experimental/join.py

```diff
@@ -243,23 +291,41 @@ def _(
 
     if _should_bcast_join(ir, left, right, partition_info, output_count):
         # Create a broadcast join
-        return _make_bcast_join(
+        joined, partition_info = _make_bcast_join(
             ir,
             output_count,
             partition_info,
+            estimated_output_rows,
             left,
             right,
         )
     else:
         # Create a hash join
-        return _make_hash_join(
+        joined, partition_info = _make_hash_join(
             ir,
             output_count,
             partition_info,
+            estimated_output_rows,
             left,
             right,
         )
 
+    if (
+        estimated_output_rows is not None
+        and max_input_rows is not None
+        and output_count
+        > (new_count := max(1, estimated_output_rows * output_count // max_input_rows))
+    ):
+        # Only repartition if estimated count suggests we should go to
+        # a smaller number of partitions
+        result = Repartition(joined.schema, joined)
+        partition_info[result] = PartitionInfo.new(
+            result, partition_info, count=new_count, table_stats=join_stats
+        )
+        return result, partition_info
+    else:
+        return joined, partition_info
```

Comment:
I think this is mostly a good thing to do. The one subtle detail that worries me is that we may induce an extra shuffle operation if we change the partition count between a subsequent join on the same columns.